[클린코드 2기 최원준] 계산기 미션

polycompo opened the pull request:
안녕하세요. 클린코드 2기 최원준 입니다.
잘 부탁드립니다


## 🎯 기능 요구사항
- [X]  2개의 숫자에 대해 덧셈이 가능하다.
- [X]  2개의 숫자에 대해 뺄셈이 가능하다.
- [X]  2개의 숫자에 대해 곱셈이 가능하다.
- [X]  2개의 숫자에 대해 나눗셈이 가능하다.
- [X]  AC(All Clear)버튼을 누르면 0으로 초기화 한다.
- [X]  숫자는 한번에 최대 3자리 수까지 입력 가능하다.
- [X]  계산 결과를 표현할 때 소수점 이하는 버림한다.


## 📊 테스트 요구사항
- [X]  기능 요구사항에 제시된 7개의 항목에 대해 테스트 케이스를 만든다.

## 문의사항
- 작성하면서 계산과 관련된 파일은 따로 분리했는데, 사용자 입력부분에 해당하는 view 부분 제외하고는 분리를 못하였습니다. 
사용자 입력부분에서 최대 입력자릿수나 연산자 연속입력시 alert을 표시할때 검증 로직이
사용자 입력에 의해 발생하는 부분이라 view에 둬도 되겠다 싶었는데 이게 맞을까요?
(검증 로직에 대한 데이터인 Total 부분이 view에서 받아오기 때문에 여기에 검증 로직을 두어야겠다 생각을 하였습니다.)

Commit:
[Setting] ESLint, Prettier 설정 추가

Commit:
[Constants] operator, calculator 상수 추가

Commit:
[Setting] Cypress 추가 및 설정

Commit:
[Setting] ESLint 설정 변경

- cypress TestCase ignore 처리
- 확장자 명시할 때 Lint 에러 예외처리
- browser window 객체 예외처
- ESLint airbnb에서 필요한 plugin 추가

Commit:
[Setting] Git ignore 수정

- idea, cypress

Commit:
[feature] 계산기 요구사항 구현

Commit:
[feature] 계산기 테스트 케이스

pocojang commented on .eslintignore:
```diff
@@ -0,0 +1,2 @@
+# Test
+cypress/
```
EOL이 계속 발생하네요~
이 문제를 한번 잡아보세요!

pocojang commented on .eslintrc.json:
```diff
@@ -0,0 +1,30 @@
+    "env": {
+        "browser": true,
+        "es2021": true
+    },
+    "ignorePatterns": [
+        "node_modules/"
+    ],
+
+    "rules": {
+        "prettier/prettier": "error",
+        "import/prefer-default-export": "off",
+        "import/extensions": [ "off" ]}
+}
```
`prettier`를 사용하셨지만 `json`은 정리되고 있지 않네요~

polycompo replied:
넵! 적용하겠습니다

pocojang commented on package.json:
```diff
@@ -0,0 +1,17 @@
+    "eslint-config-prettier": "^8.5.0",
+    "eslint-plugin-import": "^2.25.4",
+    "eslint-plugin-prettier": "^4.0.0",
+    "eslint-plugin-jsx-a11y": "^6.5.1",
+    "eslint-plugin-react": "^7.29.3",
+    "eslint-plugin-react-hooks": "^4.3.0",
```
`jsx`, `react`는 전혀 사용되지 않아서 불필요해보이네요~!

polycompo replied:
넵! 제거하겠습니다

pocojang commented on src/js/calculator.js:
```diff
@@ -0,0 +1,46 @@
+  return total / value;
+}
+
+export default function calculate(total, value, operator) {
```
매개변수를 한번에 3개 선언하여 `calculate()`를 사용하는 곳에서 그 규칙을 알고 사용하기에는 헷갈릴 것 같습니다.
인자를 순서에 맞춰서 넘겨야하는 강제성도 애매하고요.

추가적으로 인자가 아무것도 들어오지 않았을때의 기본값 대응도 고려해볼 수 있겠어요

polycompo replied:
옙. 객체로 받아볼수있도록 수정하겠습니다.

pocojang commented on src/js/calculator.js:
```diff
@@ -0,0 +1,46 @@
+  SUBTRACT_OPERATOR,
+} from './constants/operator.js';
+
+function add(total, value) {
```
`add()`라는 함수 네이밍이 어울리지는 않아보이는데 어떻게 생각하시나요~?

polycompo replied:
확실히 `add`라는 함수명은 무언가를 더한다는 느낌이 있네요. 
합산을 의미하는 `sum`으로 바꾸겠습니다

pocojang commented on src/js/calculator.js:
```diff
@@ -0,0 +1,46 @@
+}
+
+export default function calculate(total, value, operator) {
+  if (total === undefined || value === undefined || operator === undefined) {
```
방어처리는 언제나 좋습니다 👍 
하지만 방어해야할 대상 값이 늘어날때는 어떻게 확장할 수 있을까요?
또한 숫자가 아닌 것이 들어왔을때에 대한 방어가 아쉬움으로 남네요

pocojang commented on src/js/calculator.js:
```diff
@@ -0,0 +1,46 @@
+    return total;
+  }
+
+  if (operator === ADD_OPERATOR) {
+    return add(Number(total), Number(value));
+  }
+
+  if (operator === SUBTRACT_OPERATOR) {
+    return subtract(Number(total), Number(value));
+  }
+
+  if (operator === POW_OPERATOR) {
+    return pow(Number(total), Number(value));
+  }
+
+  if (operator === DIVIDE_OPERATOR) {
+    return divide(Number(total), Number(value));
+  }
```
규칙적인 매우 일정한 4가지의 분기가 약속된 경우로 보이니 `switch~case`를 고려하시는 것도 좋습니다 :)

polycompo replied:
넵 수정하겠습니다

pocojang commented on src/js/calculator.js:
```diff
@@ -0,0 +1,46 @@
+    return divide(Number(total), Number(value));
+  }
+
+  throw new Error('NOT SUPPORT OPERATOR');
```
`default case`를 에러로 던지는 좋은 시도네요 👍

pocojang commented on src/js/calculatorView.js:
```diff
@@ -0,0 +1,122 @@
+const CALCULATE_ELEMENT = document.getElementById('calculate');
+const DIGITS_ELEMENTS = document.querySelectorAll('.digit');
+const OPERATORS_ELEMENTS = document.querySelectorAll(
+  '.operation:not(#calculate)'
```
`:not`으로 셀렉팅을 하다니 멋진 시도입니다 👍

polycompo replied:
https://github.com/next-step/js-calculator/pull/92#discussion_r825265178 과 같이 이벤트 위임을 이용하여 변경해보겠습니다.

pocojang commented on src/js/constants/calculator.js:
```diff
@@ -0,0 +1,2 @@
+export const INITIAL_DIGITS = '0';
+export const MAX_INPUT_DIGITS = 3;
```
`DIGITS`을 관리하시는데 초기 값만 문자열로 관리하시는 이유가 어떻게 되시나요~?

polycompo replied:
Total 부분이 숫자로만 구성된게 아니라 연산자와 같이 구성되어 30X8과 같이 문자열로 구성되기때문에 초기값도 문자열로 처리 하였었습니다.
초기값을 숫자로 관리해보는 쪽으로 해보겠습니다.

pocojang replied:
단지 일관성이라는 포커스로 여쭤봤습니다ㅎㅎ
저는 개인적으로 일관성을 매우 중요하게 생각하는데요

그 이유는  어떠한 개발자가 보더라도 @circleGiven 님의 코드를 보면 `**_DIGITS`이라는 네이밍에서 어떠한 경우에도 일관성있는 데이터 타입을 가지는구나라는 신뢰를 주는 것이 중요합니다 :)

그렇다면 함께 네이밍 컨벤션을 세우지 않아도 암묵적인 일관성으로 코드를 읽어나가는 속도나 신뢰도가 개선되겠죠!

polycompo replied:
아하! 그렇군요
앞으로 일관성을 유지해보도록 하겠습니다.
감사합니다 ㅎ

pocojang commented on src/js/constants/operator.js:
```diff
@@ -0,0 +1,13 @@
+export const POW_OPERATOR = 'X';
+export const DIVIDE_OPERATOR = '/';
+
+export const OPERATORS = [
+  ADD_OPERATOR,
+  SUBTRACT_OPERATOR,
+  DIVIDE_OPERATOR,
+  POW_OPERATOR,
+];
```
한번 만든 상수를 잘 활용하시는 게 보기 좋네요!

pocojang commented on src/js/index.js:
```diff
@@ -0,0 +1,7 @@
+import initialize from './calculatorView.js';
+
+export default function App() {
+  initialize();
```
`initialize()`의 경우앱을 초기화하는 느낌으로 사용되지만 이벤트 바인딩만 하고 있어 그 네이밍이나 구조가 애매하게 느껴지네요

polycompo replied:
넵 이벤트 바인딩을 의미하는 이름으로 수정하겠습니다.

pocojang commented on src/js/calculatorView.js:
```diff
@@ -0,0 +1,122 @@
+}
+
+export default function initialize() {
+  // Digits Event Binding
```
불필요할 정도로 친절한 주석으로 보이네요 :)
없어도 이해에 어려움이 있지는 않을 것 같습니다

polycompo replied:
넵 제거하겠습니다

pocojang commented on src/js/calculatorView.js:
```diff
@@ -0,0 +1,122 @@
+export default function initialize() {
+  // Digits Event Binding
+  DIGITS_ELEMENTS.forEach((d) =>
+    d.addEventListener('click', () => digitHandler(d.value))
```
단수를 의미하는 느낌의 네이밍으로 `d`와 같은 축약 네이밍을 사용하신 것으로 보이네요~
조금은 귀찮더라도 의미있는 네이밍을 사용하는 것이 어떨까요?

polycompo replied:
넵 `d`를 `element`로 바꾸겠습니다

pocojang commented on src/js/calculatorView.js:
```diff
@@ -0,0 +1,122 @@
+    return;
+  }
+
+  if (isLastValueOperator() === false && isMaxDigit(getLastDigitsFromTotal())) {
```
`isLastValueOperator() === false`와
`isMaxDigit(getLastDigitsFromTotal())`이 각자하는 역할은 이해할 수 있겠는데요

`or`가 아닌 `and`의 경우 종합적으로 어떤 역할을 하는지 알려주는게 좋습니다

```js
const ??? = isLastValueOperator() === false && isMaxDigit(getLastDigitsFromTotal())

if (???) {
  notificationMaxInputDigits();
  return;
}
```

이런 활용은 어떠신지 한번 제안해봅니다 :)

polycompo replied:
넵! 의미를 부여하도록 `const`로 빼서 의미있는 네이밍을 부여하겠습니다.

pocojang commented on src/js/calculatorView.js:
```diff
@@ -0,0 +1,122 @@
+}
+
+function isLastValueOperator() {
+  return OPERATORS.some((operator) => operator === getLastValueFromTotal());
```
`some()`을 아주 적절하게 사용하고 계시네요 👍

pocojang commented on src/js/calculatorView.js:
```diff
@@ -0,0 +1,122 @@
+  return Math.floor(
+    getOperationListFromTotal().reduce((result, operator, index) => {
+      return calculate(result, getDigitListFromTotal()[index + 1], operator);
+    }, getFirstValueFromTotal())
```
`reduce()` 초기값 정의가 매우 친절하네요 👍

pocojang commented on src/js/calculatorView.js:
```diff
@@ -0,0 +1,122 @@
+}
+
+function getLastValueFromTotal() {
+  return getTotal()[getTotal().length - 1];
```
함수 내부에서 일어나는 복잡한 것을 잘 추상화해주셨네요 👍 
다만 함수가 하는 일이 늘어날 경우 `getTotal()[getTotal().length - 1];` 구문은 알아보기 힘들 수 있을 것 같네요

polycompo replied:
넵 의미를 갖도록 좀 더 추상화 하겠습니다.

pocojang commented on src/js/calculatorView.js:
```diff
@@ -0,0 +1,122 @@
+function getDigitListFromTotal() {
+  return getTotal()
+    .split(OPERATOR_REGEX)
+    .filter((v) => v);
```
`v`가 무엇을 의미하는지 명시하는 건 어떨까요?

polycompo replied:
`v`를 `operator` 로 명시하겠습니다

pocojang commented on src/js/calculatorView.js:
```diff
@@ -0,0 +1,122 @@
+function getDigitListFromTotal() {
+  return getTotal()
+    .split(OPERATOR_REGEX)
+    .filter((v) => v);
```
메서드 체이닝 활용 매우 좋습니다 👍

pocojang commented on index.html:
```diff
@@ -11,28 +11,29 @@
                 </div>
                 <div class="modifiers subgrid">
-                    <button class="modifier">AC</button>
+                    <button id="initial" class="modifier">AC</button>
```
`.modifier`만으로 충분할 것 같은데 `#initial`을 지정하신 이유가 있을까요?

polycompo replied:
지금은 `.modifier`가 하나뿐이라 충분하지만, 만약 두개 이상 들어올 경우 의도대로 동작하지 않아 `#initial`을 붙였습니다.

pocojang replied:
굉장히 엣지 케이스를 잘 고려하시는군요
저는 의미 없는 할당은 아닐까해서 여쭤봤는데 좋습니다!

pocojang commented on index.html:
```diff
@@ -11,28 +11,29 @@
-                    <button class="digit">2</button>
-                    <button class="digit">1</button>
-                    <button class="digit">0</button>
+                    <button class="digit" value="9">9</button>
+                    <button class="digit" value="8">8</button>
+                    <button class="digit" value="7">7</button>
+                    <button class="digit" value="6">6</button>
+                    <button class="digit" value="5">5</button>
+                    <button class="digit" value="4">4</button>
+                    <button class="digit" value="3">3</button>
+                    <button class="digit" value="2">2</button>
+                    <button class="digit" value="1">1</button>
+                    <button class="digit" value="0">0</button>
```
별도로 `value`를 부여하신 이유가 있으실까요~?

polycompo replied:
특별한 이유는 없었습니다. innerText를 읽어서 처리하는게 좋을까요?

pocojang replied:
그것보다는 `button`이라는 요소의 공식 명세를 한번 찾아보세요!
`HTML`을 활용하는 프론트엔드 개발자의 입장으로 생각해본다면 `HTML` 요소마다 각각의 특성을 올바르게 사용해야합니다.

때문에 단지 임의의 데이터를 HTML 요소의 속성으로 지정하고 싶다? 라는 생각이시라면 [data-*](https://developer.mozilla.org/ko/docs/Learn/HTML/Howto/Use_data_attributes)를 사용하시는게 어떨까요?

polycompo replied:
아하 넵 감사합니다!

pocojang commented on src/js/calculatorView.js:
```diff
@@ -0,0 +1,122 @@
+function calculateTotal() {
+  return Math.floor(
+    getOperationListFromTotal().reduce((result, operator, index) => {
+      return calculate(result, getDigitListFromTotal()[index + 1], operator);
```
`calculate()`만 따로 분리한 것이 매우 좋네요 👍 
`getDigitListFromTotal()[index + 1]`의 경우 무엇을 의미하는지 해당 구문에서만 보기에는 애매함이 있습니다 :)

polycompo replied:
넵. 무엇을 의미하는지 알아볼 수 있도록 수정하겠습니다.

pocojang commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,148 @@
+      checkEqualTotal(123);
+    });
+
+    it('연속적인 연산자 입력 불가', () => {
```
이런 세부적인 테스트 좋네요 👍

pocojang commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,148 @@
+const APP_URL = '../../index.html';
+const TOTAL_ID = '#total';
+
+function digitHandler(digit) {
+  return cy.get('.digit').contains(digit).click();
+}
+
+function operatorHandler(operator) {
+  return cy.get('.operation:not(#calculate)').contains(operator).click();
+}
+
+function calculateHandler() {
+  return cy.get('#calculate').click();
+}
+
+function checkEqualTotal(value) {
+  return cy.get(TOTAL_ID).should('have.text', value);
+}
```
테스트 함수 조차 반복되는 부분을 분리하셨네요 👍 
`cypress` 기능 중 동일한 역할을 해주는 것이 있으니 향후에 한번 고려해보세요  :)

pocojang commented on src/js/calculatorView.js:
```diff
@@ -0,0 +1,122 @@
+
+export default function initialize() {
+  // Digits Event Binding
+  DIGITS_ELEMENTS.forEach((d) =>
```
이벤트 위임을 활용해본다면 불필요한 이벤트 리스너 바인딩이 줄어들 수 있답니다 :)

polycompo replied:
넵! 이벤트 위임을 활용하여 리스너 바인딩을 줄여보겠습니다.

pocojang replied:
제안일 뿐이지 필수는 아닙니다 ㅎㅎ

pocojang reviewed (changes requested):
안녕하세요
원준님 반갑습니다 :)
리뷰어 장현석입니다

늦은 리뷰 죄송합니다 ㅠㅠ
계산기 미션은 온보딩 성격이 강하기때문에 그에 맞는 정도의 리뷰를 드리도록 하겠습니다.

## 구조

요구사항에 적절한 구조로 보입니다.
이런 간단한 요구사항에도 오버 엔지니어링을 하며 불필요한 코드를 생산하는 경우가 있는데
현재 구조는 정말 적당한 볼륨으로 보여서 합리적인 구조를 작성해주신 부분이 보기 좋았습니다

`src/js/calculator.js`의 경우 아주 적절하게 계산기가 동작하는 부분만 모아서 핸들링해주는 명확한 역할을 부여받은 것으로 보입니다.

`src/js/calculatorView.js` view 역할을 부여한 듯한 파일 네이밍인데 생각보다 많은 일들을 하고 있습니다.
향후 앱이 커지면 1순위로 분리되어야할 것으로 보입니다.

## 함수 활용

함수를 분리하고 추상화하는 것에 노력을 기울이신 것으로 보입니다.
저 또한 굉장히 보기 좋았고요.

하지만 너무 잘게 나뉘어진 부분이 있었습니다.
나뉘어져있지만 다른 함수와의 의존성을 강하게 지니고 있는데요.
이런 경우 분리된 효과를 보기도 힘들고 오히려 함수 실행 순서의 맥락을 알 수 없어 헷갈릴 수 있습니다.
물론 현재 단계의 요구사항 수준에서는 문제가 없지만 향후 스케일 업한 구조를 다루실때는 문제가 생길 수 있습니다.

---

일단 `Request changes`드릴테니 수정하실 부분있으시면 수정해보시고 그게 아니라면 `Approve`하도록 하겠습니다
첫 미션 고생하셨어요!

pocojang commented on .gitignore:
```diff
@@ -102,3 +102,8 @@ dist
+.idea
+
+# cypress
+cypress/fixtures/example.json
```
https://github.com/next-step/js-calculator/pull/92/files#r825261124

pocojang commented on .prettierrc:
```diff
@@ -0,0 +1,6 @@
+  "tabWidth": 2,
+  "semi": true,
+  "singleQuote": true
+}
```
https://github.com/next-step/js-calculator/pull/92/files#r825261124

pocojang commented on cypress.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "pluginsFile": false,
+  "supportFile": false
+}
```
https://github.com/next-step/js-calculator/pull/92/files#r825261124

pocojang commented on index.html:
```diff
@@ -11,28 +11,29 @@
-                    <button class="operation">-</button>
-                    <button class="operation">+</button>
-                    <button class="operation">=</button>
+                    <button class="operation" value="/">/</button>
+                    <button class="operation" value="X">X</button>
+                    <button class="operation" value="-">-</button>
+                    <button class="operation" value="+">+</button>
+                    <button class="operation" id="calculate">=</button>
```
https://github.com/next-step/js-calculator/pull/92/files#r825264114

Commit:
[Fix] eslint 설정 파일 prettier 적용

Commit:
[Fix] add 메서드 이름 sum으로 수정

Commit:
[Fix] array-util 추가

Commit:
[Fix] getDigitListFromTotal, getLastValueFromTotal, initialize 메서드 수정

- getDigitListFromTotal: filter에서 v를 유의미한 의미를 갖도록 operator로 변경
- getLastValueFromTotal: getTotal().length - 1을 추상화
- initialize: 불필요한 주석 제거

Commit:
[Fix] initialize 에서 d를 element로 수정

Commit:
[Fix] calculate 메서드 params 형식 object로 변경

Commit:
[Fix] calculateTotal, eventHandler 메서드 수정

- calculateTotal: calculate 메서드 호출시 params 형태 object로 변경 및 value를 const로 분리하여 추상화
- eventHandler: value가 아닌 event를 읽어오도록 변경

Commit:
[Fix] event binding 이벤트 위임을 이용하여 변경

- calculate operator 상수 추가
- eventBinding 쪽에서 operator로 분기처리

Commit:
[Fix] calculator 테스트 케이스 핸들러 수정

Commit:
[Fix] 불필요한 eslint plugin 제거

Commit:
[Fix] calculate 메서드 수정

- type 체크 추가
- type 체크 추가에 따른 Number 포맷팅 외부로 이동
- if 문을 case 문으로 변경

Commit:
[Fix] initialize 명칭 eventBindings 로 변경

Commit:
[Fix] digit과 operator에 data-value 추가

Commit:
[Fix] gitattributes EOL 설정 추가

Commit:
[Fix] gitattributes EOL 설정 수정

Commit:
[Fix] esintignore 주석제거

Commit:
[Fix] prettier 적용

Commit:
[Fix] gitattributes 수정

pocojang reviewed (approved):
안녕하세요
@circleGiven 

계산기 미션 완료 축하드립니다!
이번 미션은 온보딩이 목적이라 리뷰는 이만 마치도록 하겠습니다

이번에 @circleGiven 님과 피드백을 주고 받으며 느낀 점은 많은 부분 의견 수용을 하신다고 느꼈습니다 👍 
협업하기에는 누구나 좋은 동료이겠지만 그래도 이번 교육 코스에서는 자신의 의견과 생각을 주고 받으며 성장하시는 계기를 가져보시길 바랍니다 😄 

다음 미션에서는 새로운 리뷰어를 만날텐데요
앞으로도 많은 배움 얻어가세요!

polycompo commented:
> 

꼼꼼히 보시면서 리뷰해주셔서 정말 감사합니다!
